src/chatd.cpp

```diff
@@ -2667,28 +2684,121 @@ void Chat::sendSync()
     sendCommand(Command(OP_SYNC) + mChatId);
 }
 
+const Chat::PendingReactions& Chat::getPendingReactions() const
+{
+    return mPendingReactions;
+}
 
-void Chat::addReaction(const Message &message, const std::string &reaction)
+int Chat::getPendingReactionStatus(const string &reaction, Id msgId) const
 {
-    auto wptr = weakHandle();
-    marshallCall([wptr, this, message, reaction]()
+    for (auto &auxReaction : mPendingReactions)
     {
-        if (wptr.deleted())
+        if (auxReaction.mMsgId == msgId
+            && auxReaction.mReactionString == reaction)
+        {
+            return auxReaction.mStatus;
+        }
+    }
+    return -1;
+}
+
+void Chat::addPendingReaction(const std::string &reaction, const std::string &encReaction, Id msgId, uint8_t status)
+{
+    for (auto &auxReaction : mPendingReactions)
+    {
+        // If reaction already exists in pending list, only update it's status
+        if (auxReaction.mMsgId == msgId
+            && auxReaction.mReactionString == reaction)
+        {
+            auxReaction.mStatus = status;
             return;
+        }
+    }
 
-        mCrypto->reactionEncrypt(message, reaction)
-        .then([this, wptr, &message](std::shared_ptr<Buffer> data)
+    mPendingReactions.emplace_back(reaction, encReaction, msgId, status);
+}
+
+void Chat::removePendingReaction(const string &reaction, Id msgId)
+{
+    for (auto it = mPendingReactions.begin(); it != mPendingReactions.end(); it++)
+    {
+        if (it->mMsgId == msgId
+            && it->mReactionString == reaction)
         {
-            if (wptr.deleted())
-                return;
+            mPendingReactions.erase(it);
+            return;
+        }
+    }
+}
 
-           std::string encReaction (data->buf(), data->bufSize());  // lenght must be only 1 byte. passing the buffer uses 4 bytes for size
-           sendCommand(Command(OP_ADDREACTION) + mChatId + client().myHandle() + message.id() + (int8_t)data->bufSize() + encReaction);
-        })
-        .fail([this](const ::promise::Error& err)
+void Chat::retryPendingReactions()
+{
+    for (auto& reaction: mPendingReactions)
+    {
+        assert(!reaction.mReactionStringEnc.empty());
+        sendCommand(Command(reaction.mStatus) + mChatId + client().myHandle() + reaction.mMsgId + (int8_t)reaction.mReactionStringEnc.size() + reaction.mReactionStringEnc);
+    }
+}
+
+void Chat::cleanPendingReactionsByMsg(karere::Id msgid, Idx idx)
+{
+    for (auto it = mPendingReactions.begin(); it != mPendingReactions.end();)
```

Comment:
There is a very handy method for `std::list` --> `remove_if`
http://www.cplusplus.com/reference/list/list/remove_if/

We should take advantage of those algorithms when possible, since they are easier to read and usually more efficient. The implementation, in example, for a similar method:
``` cpp
void Chat::cleanPendingReactions(const karere::Id &msgId)
{
    mPendingReactions.remove_if([msgId](PendingReaction& reaction)
    {
        return reaction.mMsgId == msgId;
    });
}
```

Reply:
Done!